packages/peregrine/src/hooks/useQueryResult.js

```diff
@@ -0,0 +1,42 @@
+import { useReducer } from 'react';
+
+const initialState = {
+    data: null,
+    error: null,
+    loading: false
+};
+
+const reducer = (state, { payload, type }) => {
+    switch (type) {
+        case 'set data': {
```

Comment:
Since the `dispatch` function returned by `useReducer` is not global, these strings are not global. Actions dispatched by this function will only be handled by this reducer, so we can use extremely human-friendly action types, such as these strings.

- These strings, and this reducer, have nothing to do with Apollo; they're custom.
- I chose the simplest, most obvious names based on what the reducer does with them.
- [Compare to the docs](https://reactjs.org/docs/hooks-reference.html#usereducer): `increment`, `decrement`, `reset`.

Without all the redux boilerplate, it doesn't make as much sense to export action types, since you wouldn't import actions from a hook file. Perhaps we could return them from the hook?

```es6
const { actions, dispatch, state } = useQueryResult()
const { data, error, loading } = state

useEffect(
  () => {
    dispatch({
      payload: true,
      type: actions.setLoading
    })
  },
  []
)
```

On the other hand, exposing `dispatch` and action types for consumers to manually construct actions and dispatch them doesn't really feel right. In Venia, we send [bound action creators](https://github.com/magento-research/pwa-studio/blob/develop/packages/venia-concept/src/components/MiniCart/trigger.js#L16) to our components instead:

```es6
// 🙂 
const Trigger = props => {
  const { closeDrawer } = props

  return (
    <button onClick={props.closeDrawer}>
      Click me
    </button>
  )
}

// 😕 
const Trigger = props => {
  const { dispatch } = props
  const closeDrawer = () => {
    dispatch({
      payload: null,
      type: appActions.toggleDrawer
    })
  }

  return (
    <button onClick={closeDrawer}
      Click me
    </button>
  )
}
```

We could easily do this with hooks such as `useQueryState`. Instead of returning `dispatch` and action types, we could return functions that create actions and dispatch them. To be clear, that's actually my preferred way to do this, and it's closer to what we already do in Venia.

Reply:
> Instead of returning dispatch and action types, we could return functions that create actions and dispatch them. To be clear, that's actually my preferred way to do this, and it's closer to what we already do in Venia.

I like this idea a lot.